Add means to launch with optimum appHMIType (#880)
smartdevicelink/sdl_evolution: proposals/NNNN-Add-means-to-launch-with-optimum-appHMIType.md

# Add means to launch with optimum appHMIType

* Proposal: [SDL-NNNN](NNNN-Add-means-to-launch-with-optimum-appHMIType.md)
* Author: [Shohei Kawano](https://github.com/Shohei-Kawano)
* Status: Awaiting review
* Impacted Platforms: [ Core / iOS / Java Suite / RPC / HMI ]

## Introduction

This proposal provides a mechanism to notify SDL apps with multiple `appHMIType` which types are activated by the HMI.  

> SDLApp can display and operate according to `appHMIType` by receiving notification from HMI.  

Reviewer: ```suggestion
An SDL app can display and operate according to `appHMIType` by receiving notification from HMI.  
```

## Motivation

> Currently, in the RPC specification, SDLApp can set multiple `appHMIType`.  

Reviewer: ```suggestion
Currently, in the RPC specification, an SDL app can set multiple `appHMIType`.  
```

However, because there is no means to notify which `appHMIType` was started from HMI, it cannot be operated according to the application.  
For example, if the specification allows you to launch an SDL app with both Navigation / Media `appHMIType` by selecting an audio source,   

> the SDL app can implement a specification that allows the audio function to be used immediately.  

Reviewer: Please move this to line 18 with the rest of the sentence.

Author: commit c30b493414b647e93f20eee7c962175c108592c0 (HEAD -> Add-means-to-launch-with-optimum-appHMIType, origin/Add-means-to-launch-with-optimum-appHMIType)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Thu Dec 5 17:18:04 2019 +0900

    Remove unnecessary line breaks

In this way, it can be thought that it can contribute to the improvement of UX.  

## Proposed solution

Add a new parameter `launchAppHMIType` to the HMI / MOBILE RPC that is sent when activating the app.  

HMI API:  

Add `launchAppHMIType` to the RPC below.  
- BasicCommunication.OnAppActivated  
- SDL.ActivateApp  

```xml
     <interface name="BasicCommunication" version="2.1.0" date="2019-03-18">

> �c

Reviewer: ```suggestion
```

Author: commit a99c95d91cc0aee3ec02b2ba9c57217d9059c6ec (HEAD -> Add-means-to-launch-with-optimum-appHMIType, origin/Add-means-to-launch-with-optimum-appHMIType)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Thu Dec 5 17:21:09 2019 +0900

    Delete double-byte characters

    <function name="OnAppActivated" messagetype="notification">
        <description>Must be sent by HU system when the user clicks on app in the list of registered apps or on soft button with 'STEAL_FOCUS' action.</description>
        <param name="appID" type="Integer" mandatory="true">
            <description>ID of selected application.</description>
        </param>
        <param name="windowID" type="Integer" mandatory="false">
            <description>
               This is the unique ID assigned to the window that this RPC is intended. If this param is not included, it will be assumed that this request is specifically for the main window on the main display. See PredefinedWindows enum.
            </description>
        </param>
+       <param name="launchAppHMIType" type="AppHMIType" mandatory="false">
+           <description>
+              A parameter that instructs an app with multiple AppHMITypes to start with a specific AppHMIType
+           </description>
+       </param>
    </function>

> �c

Reviewer: ```suggestion
```

Author: commit a99c95d91cc0aee3ec02b2ba9c57217d9059c6ec (HEAD -> Add-means-to-launch-with-optimum-appHMIType, origin/Add-means-to-launch-with-optimum-appHMIType)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Thu Dec 5 17:21:09 2019 +0900

    Delete double-byte characters

    <interface name="SDL" version="1.2.0" date="2018-09-05">

> �c

Reviewer: ```suggestion
```

Author: commit a99c95d91cc0aee3ec02b2ba9c57217d9059c6ec (HEAD -> Add-means-to-launch-with-optimum-appHMIType, origin/Add-means-to-launch-with-optimum-appHMIType)
Author: 河野 祥平 <103082@ad.nexty-ele.com>
Date:   Thu Dec 5 17:21:09 2019 +0900

    Delete double-byte characters
